src/lib.rs

```diff
@@ -1,84 +1,45 @@
-use arrow::array::{AsArray, Int64Array};
+use arrow::array::AsArray;
 use arrow::compute;
-use arrow::datatypes::DataType;
+use arrow::datatypes::{DataType, Int64Type};
 use arrow_schema::ArrowError;
 use pyo3::prelude::*;
-use pyo3_arrow::error::{PyArrowError, PyArrowResult};
-use pyo3_arrow::PyArrayReader;
+use pyo3_arrow::error::PyArrowResult;
+use pyo3_arrow::PyRecordBatchReader;
 
 #[pyfunction]
-fn sum_i64_column(values: PyArrayReader, column_name: String) -> PyArrowResult<i64> {
+fn sum_i64_column(values: PyRecordBatchReader, column_name: String) -> PyArrowResult<i64> {
     let reader = values.into_reader()?;
 
-    // Validate input.
-    let field = reader.field();
-    let data_type = field.data_type();
-    let mut found_column = false;
-    match data_type {
-        DataType::Struct(fields) => {
-            for field in fields {
-                if field.name() == &column_name {
-                    let data_type = field.data_type();
-                    if data_type != &DataType::Int64 {
-                        return Err(PyArrowError::from(ArrowError::ComputeError(format!(
-                            "Expected Int64, got {data_type}"
-                        ))));
-                    }
-                    found_column = true;
-                    break;
-                }
-            }
-            if !found_column {
-                return Err(PyArrowError::from(ArrowError::ComputeError(format!(
-                    "Column '{column_name}' not found"
-                ))));
-            }
-        },
-        data_type => {
-            return Err(PyArrowError::from(ArrowError::ComputeError(format!(
-                "Expected Struct, got {data_type}"
-            ))));
-        },
-    };
-
-    // Compute sum for each chunk.
-    let mut total: i64 = 0;
-    for array_result in reader {
-        let array = array_result?;
-        let struct_array = array.as_struct();
-        let column = struct_array.column_by_name(&column_name).expect("column name already been validated");
-        let primitive_array: &Int64Array = column.as_primitive();
-        if let Some(sum) = compute::sum(primitive_array) {
-            total += sum;
-        };
+    let schema = reader.schema();
+    let (column_index, field) =
+        schema
+            .column_with_name(&column_name)
+            .ok_or(ArrowError::ComputeError(format!(
+                "Column '{column_name}' not found"
+            )))?;
+
+    if !matches!(field.data_type(), DataType::Int8) {
```

Comment:
Oops 
```suggestion
    if !matches!(field.data_type(), DataType::Int64) {
```

Reply:
FWIW if you wanted to write more generic code, you could use [`downcast_primitive_array`](https://docs.rs/arrow/latest/arrow/macro.downcast_primitive_array.html) to support more than just Int64 typed-input.

You can look at my generic implementation in arro3-compute: https://github.com/kylebarron/arro3/blob/d430963565aea5dc9871870012ccf1fffc946a32/arro3-compute/src/aggregate.rs#L210-L252